src/view.c

```diff
@@ -19,6 +20,25 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
+struct view *
+view_from_wlr_surface(struct wlr_surface *surface)
+{
+	assert(surface);
+	if (wlr_surface_is_xdg_surface(surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		return xdg_surface->data;
+	}
+#if HAVE_XWAYLAND
+	if (wlr_surface_is_xwayland_surface(surface)) {
+		struct wlr_xwayland_surface *xsurface =
+			wlr_xwayland_surface_from_wlr_surface(surface);
+		return xsurface->data;
+	}
+#endif
+	return NULL;
+}
```

Comment:
This feels like it should be in `view-impl-common.c` instead.
IMHO we should strive to keep `view.c` clean of xdg and xwayland related stuff.

AFAIR `wlr_{xdg,xwayland}_surface_from_wlr_surface()` can return `NULL` as well so we should handle that.

Reply:
view-impl-common.c is "common code for shell `view->impl` functions" according to its own comment. This isn't called by (nor calls) any `view->impl` functions, so I'm not sure it belongs there.

It's unfortunate that this needs to do anything xdg- or xwayland-specific in the first place, since neither `view` nor `wlr_surface` are xdg- or xwayland-specific. Not sure of how to improve that at the moment though.

Reply:
Regarding returning `NULL` - you may be thinking of the `try_from` versions in wlroots 0.17. The `from` versions (without the `try`) in wlroots 0.16 never return `NULL` from what I can see, but will `assert(false)` if passed a `wlr_surface` with the wrong role. This function will look different with wlroots 0.17.

Reply:
One option to avoid all the `xdg` and `xwayland` stuff would be to loop over all the views like `desktop_focused_view()` did before. Slightly less efficient, but more generic.

Reply:
AFAIR they only changed the name of the functions, they could always return `NULL`:

https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/include/wlr/types/wlr_xdg_shell.h#L488
https://gitlab.freedesktop.org/wlroots/wlroots/-/blob/0.16/include/wlr/xwayland/xwayland.h?ref_type=heads#L231

Reply:
> I don't remember if we have to loop up the wlr_surface tree until we find the main toplevel surface.

I am not sure either. The old `desktop_focused_view()` didn't loop to find the toplevel, but that doesn't mean it was correct.

Reply:
> AFAIR they only changed the name of the functions, they could always return `NULL`

Okay, I'll add a `NULL` check

Reply:
I added `NULL` checks and TODOs regarding the other comments.